### PR TITLE
[cxx] Make some code valid C++, esp. adding casts and esp. not touching allocators.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 FullAOT

File: mono/metadata/marshal-ilgen.c [hunk 1]
```diff
@@ -560,6 +560,7 @@ emit_ptr_to_object_conv (MonoMethodBuilder *mb, MonoType *type, MonoMarshalConv
 static gpointer
 conv_to_icall (MonoMarshalConv conv, int *ind_store_type)
 {
+#define return return (gpointer)
```
Comment (luhenry) [inline mono/metadata/marshal-ilgen.c]: Remove that.

File: mono/metadata/method-builder-ilgen.h [hunk 2]
```diff
@@ -50,6 +50,8 @@ mono_mb_emit_managed_call (MonoMethodBuilder *mb, MonoMethod *method, MonoMethod
 void
 mono_mb_emit_icall (MonoMethodBuilder *mb, gpointer func);
 
+#define mono_mb_emit_icall(mb, function) mono_mb_emit_icall ((mb), (gpointer)(function))
```
Comment (luhenry) [inline mono/metadata/method-builder-ilgen.h]: Remove that.

File: mono/metadata/remoting.c [hunk 3]
```diff
@@ -108,6 +108,9 @@ register_icall (gpointer func, const char *name, const char *sigstr, gboolean sa
 	mono_register_jit_icall (func, name, sig, save);
 }
 
+#define register_icall(func, name, sigstr, save) \
```
Comment (luhenry) [inline mono/metadata/remoting.c]: Remove that.

File: mono/metadata/sgen-mono.c [hunk 4]
```diff
@@ -995,6 +995,8 @@ create_allocator (int atype, ManagedAllocatorVariant variant)
 	int num_params, i;
 
 	if (!registered) {
+// Cast the first parameter to gpointer; macros do not recurse.
+#define mono_register_jit_icall(func, name, sig, no_wrapper) (mono_register_jit_icall ((gpointer)(func), (name), (sig), (no_wrapper)))
```
Comment (luhenry) [inline mono/metadata/sgen-mono.c]: Remove that.

File: mono/metadata/threads-types.h [hunk 5]
```diff
@@ -44,6 +44,7 @@ typedef enum {
 } MonoThreadApartmentState;
 
 typedef enum {
+// These values match Windows, but are offset by 2.
```
Comment (luhenry) [inline mono/metadata/threads-types.h]: Unrelated to the change, this is to match the managed values, nothing to do with Win32. Remove that.

File: mono/metadata/threads.c [hunk 6]
```diff
@@ -1873,15 +1873,16 @@ ves_icall_System_Threading_Thread_GetPriority (MonoThreadObjectHandle this_obj,
  *
  * Sets the priority of the given thread.
  */
+ICALL_EXPORT
 void
-ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj, int priority, MonoError *error)
+ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj, int/*MonoThreadPriority FIXMEcxx*/ priority, MonoError *error)
```
Comment (luhenry) [inline mono/metadata/threads.c]: Remove this unreadable comment.

File: mono/mini/aot-runtime.h [hunk 7]
```diff
@@ -244,6 +244,7 @@ void     mono_aot_set_make_unreadable       (gboolean unreadable);
 gboolean mono_aot_is_pagefault              (void *ptr);
 void     mono_aot_handle_pagefault          (void *ptr);
 void     mono_aot_register_jit_icall        (const char *name, gpointer addr);
+#define mono_aot_register_jit_icall(name, addr) (mono_aot_register_jit_icall ((name), (gpointer)(addr)))
```
Comment (luhenry) [inline mono/mini/aot-runtime.h]: Remove that.

File: mono/mini/method-to-ir.c [hunk 8]
```diff
@@ -82,6 +82,8 @@
 #include "mini-llvm.h"
 #include "mini-runtime.h"
 
+#define mono_emit_jit_icall(a, b, c) (mono_emit_jit_icall ((a), (gconstpointer)(b), (c)))
```
Comment (luhenry) [inline mono/mini/method-to-ir.c]: Remove that.

File: mono/metadata/marshal-ilgen.c (hunk 1 above)
Comment (jaykrell) [reply mono/metadata/marshal-ilgen.c]: Understood. It reduced line damage.

File: mono/metadata/method-builder-ilgen.h (hunk 2 above)
Comment (jaykrell) [reply mono/metadata/method-builder-ilgen.h]: It is the explicit cast vs. auto cast thing.

File: mono/metadata/remoting.c (hunk 3 above)
Comment (jaykrell) [reply mono/metadata/remoting.c]: Understood. it is a way to cut a lot of line damage. I assume you understand that and will proceed with the alternative.

File: mono/metadata/sgen-mono.c (hunk 4 above)
Comment (jaykrell) [reply mono/metadata/sgen-mono.c]: Ditto.

File: mono/metadata/threads-types.h (hunk 5 above)
Comment (jaykrell) [reply mono/metadata/threads-types.h]: It is both. We subtract 2 elsewhere because of this.

File: mono/metadata/threads.c (hunk 6 above)
Comment (jaykrell) [reply mono/metadata/threads.c]: Move it? There are a few places where we have ints that are really enums.

File: mono/metadata/remoting.c (hunk 3 above)
Comment (jaykrell) [reply mono/metadata/remoting.c]: About 12 lines of "damage" here.

File: mono/metadata/sgen-mono.c (hunk 4 above)
Comment (jaykrell) [reply mono/metadata/sgen-mono.c]: About 5 lines of "damage" here.

File: mono/metadata/method-builder-ilgen.h (hunk 2 above)
Comment (jaykrell) [reply mono/metadata/method-builder-ilgen.h]: Damage could be quite high here, like 100+. I'll try it.

Comment (jaykrell) [reply mono/metadata/method-builder-ilgen.h]: Maybe an inline template instead, with a cast therein? Kinda the same as the macro really, except no need to undef or put parens around the real function. Anyway I'm adding the casts, lots of them.

File: mono/mini/aot-runtime.h (hunk 7 above)
Comment (jaykrell) [reply mono/mini/aot-runtime.h]: About 15 lines of "damage". I'll add the casts.

File: mono/mini/method-to-ir.c (hunk 8 above)
Comment (jaykrell) [reply mono/mini/method-to-ir.c]: Around 80 lines of damage.

**Comment by jaykrell** [comment]
I not super thrilled with the level of line damage here. But I agree it has obvious correct meaning.
The macros feel at about the same level of hack/engineering as a template, just less general.
Maybe try a template instead? It should work, be deemed more elegant, and avoid line damage just like the macros?

```
#include <stdio.h>

void foo(void*)
{
	printf("1\n");
}

template <typename T>
void foo(T a)
{
	printf("2\n");
	foo((void*)a);
}

int main()
{
int i;
foo(&i);
foo((void*)&i);
}
```

**Comment by jaykrell** [comment]
Redundant with https://github.com/mono/mono/pull/10252.

**Pull request closed without merging**